lib/perl/Genome/Model/Tools/Transcriptome/ErccMapUnaligned.pm

```diff
@@ -0,0 +1,367 @@
+    );
+
+    Genome::Sys->shellcmd(
+        cmd => $cmd,
+        input_files => ["$bam"],
+        output_files => ["$index_file"],
+        skip_if_output_is_present => 0
+    );
+}
+
+sub generate_idxstats {
+    my ($self, $bam) = @_;
+
+    my $samtools_idxstats_path = Path::Class::File->new(Genome::Sys->create_temp_file_path());
+
+    my $cmd = join(' ',
+        $self->samtools, 'idxstats',
+        "$bam", 
+        '>', "$samtools_idxstats_path"
+    );
+
+    Genome::Sys->shellcmd(
+        cmd => $cmd,
+        input_files => ["$bam"],
+        output_files => ["$samtools_idxstats_path"],
+        skip_if_output_is_present => 0
+    );
+
+    my $idxstats_hash_ref = Genome::Model::Tools::Sam::Idxstats->parse_file_into_hashref(
+        "$samtools_idxstats_path"
+    );
+
+    return $idxstats_hash_ref;
+}
+
+sub generate_tsvfile {
+    my ($self, $idxstats) = @_;
+
+    my $r_input_file = Genome::Sys->create_temp_file_path();
+
+    my @headers = (
+        'Re-sort ID',
+        'ERCC ID',
+        'subgroup',
+        'ERCC Mix',
+        'concentration (attomoles/ul)',
+        'label',
+        'count',
+    );
+
+    my $writer = Genome::Utility::IO::SeparatedValueWriter->create(
+        output => $r_input_file,
+        headers => \@headers,
+        separator => "\t",
+    );
+
+    my %ercc_control = $self->load_ercc_control_info;
+
+    for my $chr (keys %{$idxstats}) {
+        my $ercc_data = $ercc_control{$chr};
+        unless ($ercc_data) {
+            die('Missing chromosome: '. $chr);
+        }
+        my $concentration = $ercc_data->{'concentration in Mix 1 (attomoles/ul)'};
+        if ($self->ercc_spike_in_mix == 2) {
+            $concentration = $ercc_data->{'concentration in Mix 2 (attomoles/ul)'};
+        }
+        my %data = (
+            'Re-sort ID' => $ercc_data->{'Re-sort ID'},
+            'ERCC ID' => $ercc_data->{'ERCC ID'},
+            'subgroup' => $ercc_data->{'subgroup'},
+            'ERCC Mix' => $self->ercc_spike_in_mix,
+            'concentration (attomoles/ul)' => $concentration,
+            'label' => 'na',
+            'count' => $idxstats->{$chr}{map_read_ct},
+        );
+        $writer->write_one(\%data);
+    }
+    $writer->output->close;
+
+    return Path::Class::File->new($r_input_file);
+}
+
+sub copy_to {
+    my $self = shift;
+    my ($src, $dest) = @_;
+    if (UNIVERSAL::isa($dest, Path::Class::File::)) {
+        $dest = $dest->stringify;
+        die "Can't copy to file $dest: it is a directory" if -d $dest;
+    }
+    elsif (UNIVERSAL::isa($dest, Path::Class::Dir::)) {
+        $dest = $dest->stringify;
+        die "Can't copy to directory $dest: it is a file" if -f $dest;
+        die "Can't copy to directory $dest: no such directory"
+          unless -d $dest;
+    }
+    elsif (ref $dest) {
+        die "Don't know how to copy files to objects of type '"
+            . ref($src) . "'";
+    }
+
+    File::Copy::cp($src->stringify, "${dest}")
```

Comment:
You have to be careful with permissions here.  For example, if you are copying a temp file the permissions are usually locked down and you usually don't want permissions preserved.  `Genome::Sys` has `copy_file`, `move_file`, `rename`, etc.

> `cp` will preserve the source file's permission bits like the shell utility `cp(1)` would do, while `copy` uses the default permissions for the target file (which may depend on the process' umask, file ownership, inherited ACLs, etc.)